pkg/scheduler/schedulercache/cache.go

```diff
@@ -112,6 +112,8 @@ func (cache *schedulerCache) UpdateNodeNameToInfoMap(nodeNameToInfo map[string]*
 	cache.mu.Lock()
 	defer cache.mu.Unlock()
 	for name, info := range cache.nodes {
+		// Transient scheduler info is added here.
+		info.TransientInfo = newTransientSchedulerInfo()
```

Comment:
Please check the flag gate here too. I wonder how much is impact of this on performance, particularly the impact of allocating/garbage collecting memory for all the nodes in every scheduling cycle. I would initialize globally once (may be with the NodeInfo initialization) and I would just clear it here to avoid frequent allocation/garbage collection.

Reply:
> I wonder how much is impact of this on performance, particularly the impact of allocating/garbage collecting memory for all the nodes in every scheduling cycle

I had a similar concern so I added the comment saying that add fields to the transient structure only if necessary. I now added a benchmark for 1knode and 30 pods on each node for updateNodeNameToInfo function. 

But for the following run, I changed the node count to 10k with 30 pods on each node to stress it further.

Without transient info structure.
```bash
 /usr/local/go/bin/go test -v k8s.io/kubernetes/pkg/scheduler/schedulercache -bench ^BenchmarkUpdate1kNodes30kPods$ -run=^$ -cpuprofile=cpu_with10knodeswithout.out
goos: linux
goarch: amd64
pkg: k8s.io/kubernetes/pkg/scheduler/schedulercache
BenchmarkUpdate1kNodes30kPods-8   	     100	  16567535 ns/op
PASS
ok  	k8s.io/kubernetes/pkg/scheduler/schedulercache	5.613s
```

With transient info structure.
```
/usr/local/go/bin/go test -v k8s.io/kubernetes/pkg/scheduler/schedulercache -bench ^BenchmarkUpdate1kNodes30kPods$ -run=^$ -cpuprofile=cpu_with10knodeswith.out
goos: linux
goarch: amd64
pkg: k8s.io/kubernetes/pkg/scheduler/schedulercache
BenchmarkUpdate1kNodes30kPods-8   	     100	  16487893 ns/op
PASS
ok  	k8s.io/kubernetes/pkg/scheduler/schedulercache	5.601s
```

So, there is not much of difference between both the runs. Not sure if my analysis is wrong but it could be because of these structures being allocated on stack as pointers are being passed around. I am also attaching call graph pdfs with and without transient info struct allocation. The functions I have looked at include runtime's malloc.go and runtime's mgcmark.go

[cpu_with10knodeswith.pdf](https://github.com/kubernetes/kubernetes/files/1845004/cpu_with10knodeswith.pdf)
[cpu_with10knodeswithout.pdf](https://github.com/kubernetes/kubernetes/files/1845005/cpu_with10knodeswithout.pdf)


> I would initialize globally once (may be with the NodeInfo initialization) and I would just clear it here to avoid frequent allocation/garbage collection.

The problem is we are calling updateNodeNameToInfo function at the start of every scheduling cycle and we need to initialize these values everytime. Also, I am not sure what clearing them means in the context of garbage collection, resetting the needed values to 0?

Reply:
Thanks for your analysis. Looks like adding the transient info struct has not had much impact on performance.

> it could be because of these structures being allocated on stack as pointers are being passed around.

I doubt if they are stack allocated when built with the rest of the code. The pointer escapes the allocating function, but when you run `UpdateNodeNameToInfoMap` by itself in a tight loop, then there is chance that the pointer is stack allocated. That could be the reason that the benchmark numbers are so close.

> Also, I am not sure what clearing them means in the context of garbage collection, resetting the needed values to 0?

Yes. I meant setting the struct members to an invalid value when clearing them. The invalid value could be 0 or a negative value that they never get when they are used.

Reply:
> The pointer escapes the allocating function, but when you run UpdateNodeNameToInfoMap by itself in a tight loop, then there is chance that the pointer is stack allocated.

Thanks. That makes sense too..

> Yes. I meant setting the struct members to an invalid value when clearing them. The invalid value could be 0 or a negative value that they never get when they are used.

I now added a reset function which resets the value to 0 instead of allocating new structs again. PTAL.